recipe/meta.yaml

```diff
@@ -12,16 +12,17 @@ source:
   sha256: {{ sha256 }}
 
 build:
-  number: 0
+  number: 1
 
 requirements:
   build:
-    - toolchain
     - cmake
-    - zeromq 4.2.3
+    - {{ compiler('c') }}
+  host:
+    - zeromq
 
   run:
-    - zeromq 4.2.3
+    - zeromq
```

Comment:
Looking at the cppzmq github, it looks like a given version should work with more than one version of zeromq, but I'm no expert. The zeromq version is pinned globally in conda-forge with conda-forge-pinnings, and that also puts an 'x.x' pin on the runtime requirement to match the build version up to the minor version number. So I think dropping the version number from the run requirement should be fine, unless you want a tighter pinning than 'x.x'. This way, you can just do a 'conda-smithy rerender' to see if a new build is needed to update the pins with to the global conda-forge values.

Reply:
So my problem is that I have a package I'm working on (gnuradio) which requires cppzmq, which right now requires zeromq 4.2.3. But the pinned version in conda-forge currently is 4.2.5, so I'm running into a conflict (re-thinking it now, maybe I can just depend on cppzmq which I can count on to pull in zeromq and guarantee no conflicts; that still depends on a user not wanting to install gnuradio with anything that requires the newer zeromq). But it's not like this is just an old version of cppzmq; 4.2.3 is the current version, and it seems like there's not a newer version because it's intended to still work with zeromq 4.2.5. But I also get the desire to limit headaches by more closely linking the cppzmq and zeromq versions than what may strictly be necessary, and I certainly don't know the best way to handle it.

Reply:
I would prefer then relationship between zeromq and cppzmq to be looser...

Reply:
> I am not an expert about the addition of pinning, but that seems like a weird choice to pin a package like zeromq globally

conda-forge has global *default* pins for packages, to help ensure compatibility across packages that use the same dependency. The current pinning for zeromq:

```yaml
zeromq: 4.2.5
pin_run_as_build:
  zeromq:
    max_pin: x.x
```

That means that all packages that link libzmq will *build* with 4.2.5 and will have a runtime dependency on `>=4.2.5,<4.3`. Assuming we finished working out the ABI/SONAME issues in libzmq (I think we have), this should be correct and safe. Any package actually built and linked with 4.2.5 is *not* compatible with 4.2.3 and earlier (this is an issue with how we built the conda package for earlier versions, not an issue with libzmq versions themselves).

Since cppzmq is header-only, it can presumably have a looser dependency on zeromq, and specify that it should work with any 4.2.x (or `>=4.2.3,<4.3`) instead of pinning to 4.2.3. It's fine to do this in meta.yaml, but it can also do this in `recipe/conda_forge_config.yaml`.

Reply:
The thing is that the author of `cppzmq` numbers cppzmq exactly in the same way as `zeromq`, sort of implying that the two are pinned together.

I confirm that in practice, cppzmq works with multiple versions of zeromq, but I don't know what the intent is.

cc @bluca do you plan on having different version numbers for cppzmq and zeromq?

Reply:
Note that my preference would have been the two package to have completely different version number and cppzmq to specify a range for its dependency on zeromq.

cf https://github.com/zeromq/cppzmq/issues/108

Reply:
I don't mind either way, and there is no compatibility requirement - apart from the obvious: if a wrapper is for an API added in $version, it will not be available with $version - 1. Which is why I think that scheme was used - that way you know immediately what the wrappers API full compatibility list is